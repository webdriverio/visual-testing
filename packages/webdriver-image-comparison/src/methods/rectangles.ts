import type { ChainablePromiseElement } from 'webdriverio'
import { calculateDprData, getBase64ScreenshotSize, isObject } from '../helpers/utils.js'
import { getElementPositionAndroid, getElementPositionDesktop, getElementWebviewPosition } from './elementPosition.js'
import type {
    DeviceRectangles,
    ElementRectangles,
    RectanglesOutput,
    ScreenRectanglesOptions,
    StatusAddressToolBarRectangles,
    StatusAddressToolBarRectanglesOptions,
} from './rectangles.interfaces.js'
import type { GetElementRect } from './methods.interfaces.js'
import type { CheckScreenMethodOptions } from '../commands/screen.interfaces.js'
import type { InstanceData } from './instanceData.interfaces.js'

/**
 * Determine the element rectangles on the page / screenshot
 */
export async function determineElementRectangles({
    executor,
    base64Image,
    options,
    element,
}: ElementRectangles): Promise<RectanglesOutput> {
    // Determine screenshot data
    const {
        devicePixelRatio,
        deviceRectangles,
        initialDevicePixelRatio,
        innerHeight,
        isAndroid,
        isAndroidNativeWebScreenshot,
        isEmulated,
        isIOS,
    } = options
    const internalDpr = isEmulated ? initialDevicePixelRatio : devicePixelRatio
    const { height } = getBase64ScreenshotSize(base64Image, internalDpr)
    let elementPosition

    // Determine the element position on the screenshot
    if (isIOS) {
        elementPosition = await getElementWebviewPosition(executor, element, { deviceRectangles })
    } else if (isAndroid) {
        elementPosition = await getElementPositionAndroid(executor, element, { deviceRectangles, isAndroidNativeWebScreenshot })
    } else {
        elementPosition = await getElementPositionDesktop(executor, element, { innerHeight, screenshotHeight: height })
    }
    // Validate if the element is visible
    if (elementPosition.height === 0 || elementPosition.width === 0) {
        let selectorMessage = ' '
        if (element.selector) {
            selectorMessage = `, with selector "$(${element.selector})",`
        }
        const message = `The element${selectorMessage}is not visible. The dimensions are ${elementPosition.width}x${elementPosition.height}`
        throw new Error(message)
    }

    // Determine the rectangles based on the device pixel ratio
    return calculateDprData(
        {
            height: elementPosition.height,
            width: elementPosition.width,
            x: elementPosition.x,
            y: elementPosition.y,
        },
        internalDpr,
    )
}

/**
 * Determine the rectangles of the screen for the screenshot
 */
export function determineScreenRectangles(base64Image: string, options: ScreenRectanglesOptions): RectanglesOutput {
    // Determine screenshot data
    const {
        devicePixelRatio,
        enableLegacyScreenshotMethod,
        initialDevicePixelRatio,
        innerHeight,
        innerWidth,
        isEmulated,
        isIOS,
        isAndroidChromeDriverScreenshot,
        isAndroidNativeWebScreenshot,
        isLandscape,
    } = options

    // For #967: When a screenshot of an emulated device is taken, but the browser was initially
    // started as a "desktop" session, so not with emulated caps, we need to store the initial
    // devicePixelRatio when we take a screenshot and enableLegacyScreenshotMethod is enabled
    const internalDpr = isEmulated && enableLegacyScreenshotMethod ? initialDevicePixelRatio : devicePixelRatio
    const { height, width } = getBase64ScreenshotSize(base64Image, internalDpr)

    // Determine the width
    const screenshotWidth = isIOS || isAndroidChromeDriverScreenshot ? width : innerWidth
    const screenshotHeight = isIOS || isAndroidNativeWebScreenshot ? height : innerHeight
    const isRotated = isLandscape && height > width

    // Determine the rectangles
    return calculateDprData(
        {
            height: isRotated ? screenshotWidth : screenshotHeight,
            width: isRotated ? screenshotHeight : screenshotWidth,
            x: 0,
            y: 0,
        },
        internalDpr,
    )
}

/**
 * Determine the rectangles for the mobile devices
 */
export function determineStatusAddressToolBarRectangles({ deviceRectangles, options }:{
    deviceRectangles: DeviceRectangles,
    options: StatusAddressToolBarRectanglesOptions,
}): StatusAddressToolBarRectangles {
    const {
        blockOutSideBar,
        blockOutStatusBar,
        blockOutToolBar,
        isAndroid,
        isAndroidNativeWebScreenshot,
        isMobile,
        isViewPortScreenshot,
    } = options
    const rectangles = []

    if (
        isViewPortScreenshot &&
        isMobile &&
        ( isAndroid && isAndroidNativeWebScreenshot || !isAndroid )
    ) {
        const statusAddressBar = {
            x: deviceRectangles.statusBarAndAddressBar.x, y: deviceRectangles.statusBarAndAddressBar.y,
            width: deviceRectangles.statusBarAndAddressBar.width, height: deviceRectangles.statusBarAndAddressBar.height,
        }
        const toolBar = {
            x: deviceRectangles.bottomBar.x, y: deviceRectangles.bottomBar.y,
            width: deviceRectangles.bottomBar.width, height: deviceRectangles.bottomBar.height,
        }
        const leftSidePadding = {
            x: deviceRectangles.leftSidePadding.x, y: deviceRectangles.leftSidePadding.y,
            width: deviceRectangles.leftSidePadding.width, height: deviceRectangles.leftSidePadding.height,
        }
        const rightSidePadding = {
            x: deviceRectangles.rightSidePadding.x, y: deviceRectangles.rightSidePadding.y,
            width: deviceRectangles.rightSidePadding.width, height: deviceRectangles.rightSidePadding.height,
        }

        if (blockOutStatusBar) {
            rectangles.push(statusAddressBar)
        }

        if (blockOutToolBar) {
            rectangles.push(toolBar)
        }

        if (blockOutSideBar) {
            rectangles.push(leftSidePadding, rightSidePadding)
        }
    }

    return rectangles
}

/**
 * Validate that the element is a WebdriverIO element
 */
export function isWdioElement(x: unknown) {
    if (!isObject(x)) {
        return false
    }

    const region = x as WebdriverIO.Element
    const keys: (keyof WebdriverIO.Element)[] = ['selector', 'elementId']

    return keys.every(key => typeof region[key] === 'string')
}

/**
 * Validate that the object is a valid ignore region
 */
function validateIgnoreRegion(x: unknown) {
    if (!isObject(x)) {
        return false
    }

    const region = x as RectanglesOutput
    const keys: (keyof RectanglesOutput)[] = ['height', 'width', 'x', 'y']

    return keys.every(key => typeof region[key] === 'number')
}

/**
 * Format the error message
 */
function formatErrorMessage(item:unknown, message:string) {
    const formattedItem = isObject(item) ? JSON.stringify(item) : item
    return `${formattedItem} ${message}`
}

/**
 * Split the ignores into elements and regions and throw an error if
 * an element is not a valid WebdriverIO element/region
 */
function splitIgnores(items:unknown[]): { elements: WebdriverIO.Element[], regions: RectanglesOutput[] }{
    const elements = []
    const regions = []
    const errorMessages = []

    for (const item of items) {
        if (Array.isArray(item)) {
            for (const nestedItem of item) {
                if (!isWdioElement(nestedItem)) {
                    errorMessages.push(formatErrorMessage(nestedItem, 'is not a valid WebdriverIO element'))
                } else {
                    elements.push(nestedItem as WebdriverIO.Element)
                }
            }
        } else if (isWdioElement(item)) {
            elements.push(item as WebdriverIO.Element)
        } else if (validateIgnoreRegion(item)) {
            regions.push(item as RectanglesOutput)
        } else {
            errorMessages.push(formatErrorMessage(item, 'is not a valid WebdriverIO element or region'))
        }
    }

    if (errorMessages.length > 0) {
        throw new Error('Invalid elements or regions: ' + errorMessages.join(', '))
    }

    return { elements, regions }
}

/**
 * Get the regions from the elements
 */
async function getRegionsFromElements(
    elements: WebdriverIO.Element[],
    getElementRect: GetElementRect,
): Promise<RectanglesOutput[]> {
    const regions = []
    for (const element of elements) {
        const region = await getElementRect(element.elementId)
        regions.push(region)
    }

    return regions
}

/**
 * Translate ignores to regions
 */
export async function determineIgnoreRegions(
    ignores: (RectanglesOutput | WebdriverIO.Element | ChainablePromiseElement)[],
    getElementRect: GetElementRect,
): Promise<RectanglesOutput[]>{
    const awaitedIgnores = await Promise.all(ignores)
    const { elements, regions } = splitIgnores(awaitedIgnores)
    const regionsFromElements = await getRegionsFromElements(elements, getElementRect)

    return [...regions, ...regionsFromElements]
        .map((region:RectanglesOutput) => ({
            x: Math.round(region.x),
            y: Math.round(region.y),
            width: Math.round(region.width),
            height: Math.round(region.height),
        }))
}

/**
 * Determine the device block outs
 */
export async function determineDeviceBlockOuts({ isAndroid, screenCompareOptions, instanceData }: {
    isAndroid: boolean,
    screenCompareOptions: CheckScreenMethodOptions,
    instanceData: InstanceData,
}){
    const rectangles: RectanglesOutput[] = []
    const { blockOutStatusBar, blockOutToolBar } = screenCompareOptions
    const { deviceRectangles:{ homeBar, statusBar } } = instanceData

    if (blockOutStatusBar){
        rectangles.push(statusBar)
    }
    if (isAndroid){
        //
    } else if (blockOutToolBar){
        rectangles.push(homeBar)
    }

    // @TODO: This is from the native-app-compare module, I can't really find the diffs between the two
    // if (options.blockOutStatusBar) {
    //     rectangles.push(deviceInfo.rectangles.statusBar)
    // }

    // if (driver.isAndroid && options.blockOutNavigationBar) {
    //     rectangles.push(deviceInfo.rectangles.androidNavigationBar)
    // }

    return rectangles
}
