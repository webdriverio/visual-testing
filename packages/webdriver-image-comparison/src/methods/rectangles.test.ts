import { describe, it, expect, vi } from 'vitest'
import { determineElementRectangles, determineScreenRectangles, determineStatusAddressToolBarRectangles } from './rectangles.js'
import { IMAGE_STRING } from '../mocks/mocks.js'

describe('rectangles', () => {
    describe('determineElementRectangles', () => {
        it('should determine them for iOS', async () => {
            const options = {
                isAndroid: false,
                devicePixelRatio: 2,
                deviceRectangles: {
                    bottomBar: { y: 0, x: 0, width: 0, height: 0 },
                    homeBar: { y: 0, x: 0, width: 0, height: 0 },
                    leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                    rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                    screenSize: { height: 0, width: 0 },
                    statusBar: { y: 0, x: 0, width: 0, height: 0 },
                    statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
                    viewport: { y: 20, x: 30, width: 0, height: 0 },
                },
                isAndroidNativeWebScreenshot: false,
                innerHeight: 678,
                isIOS: true,
            }
            const MOCKED_EXECUTOR = vi
                .fn()
                // getBoundingClientRect
                .mockResolvedValueOnce({
                    height: 120,
                    width: 120,
                    x: 100,
                    y: 10,
                })

            expect(
                await determineElementRectangles({
                    executor: MOCKED_EXECUTOR,
                    base64Image: IMAGE_STRING,
                    options,
                    element: 'element',
                }),
            ).toMatchSnapshot()
        })

        it('should determine them for Android Native webscreenshot', async () => {
            const options = {
                isAndroid: true,
                devicePixelRatio: 3,
                deviceRectangles: {
                    homeBar: { y: 0, x: 0, width: 0, height: 0 },
                    statusBar: { y: 0, x: 0, width: 0, height: 0 },
                    screenSize: { height: 0, width: 0 },
                    statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
                    viewport: { y: 200, x: 300, width: 0, height: 0 },
                    bottomBar: { y: 0, x: 0, width: 0, height: 0 },
                    leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                    rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                },
                isAndroidNativeWebScreenshot: true,
                innerHeight: 678,
                isIOS: false,
            }
            const MOCKED_EXECUTOR = vi
                .fn()
                // getBoundingClientRect
                .mockResolvedValueOnce({
                    height: 300,
                    width: 200,
                    x: 100,
                    y: 10,
                })

            expect(
                await determineElementRectangles({
                    executor: MOCKED_EXECUTOR,
                    base64Image: IMAGE_STRING,
                    options,
                    element: 'element',
                }),
            ).toMatchSnapshot()
        })

        it('should determine them for Android ChromeDriver', async () => {
            const options = {
                isAndroid: true,
                devicePixelRatio: 1,
                deviceRectangles: {
                    homeBar: { y: 0, x: 0, width: 0, height: 0 },
                    statusBar: { y: 0, x: 0, width: 0, height: 0 },
                    screenSize: { height: 0, width: 0 },
                    statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
                    viewport: { y: 200, x: 300, width: 0, height: 0 },
                    bottomBar: { y: 0, x: 0, width: 0, height: 0 },
                    leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                    rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                },
                isAndroidNativeWebScreenshot: false,
                innerHeight: 678,
                isIOS: false,
            }
            const MOCKED_EXECUTOR = vi
                .fn()
                // getBoundingClientRect
                .mockResolvedValueOnce({
                    height: 20,
                    width: 375,
                    x: 0,
                    y: 0,
                })

            expect(
                await determineElementRectangles({
                    executor: MOCKED_EXECUTOR,
                    base64Image: IMAGE_STRING,
                    options,
                    element: 'element',
                }),
            ).toMatchSnapshot()
        })

        it('should determine them for a desktop browser', async () => {
            const options = {
                isAndroid: false,
                devicePixelRatio: 2,
                deviceRectangles: {
                    homeBar: { y: 0, x: 0, width: 0, height: 0 },
                    statusBar: { y: 0, x: 0, width: 0, height: 0 },
                    screenSize: { height: 0, width: 0 },
                    statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
                    viewport: { y: 0, x: 0, width: 0, height: 0 },
                    bottomBar: { y: 0, x: 0, width: 0, height: 0 },
                    leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                    rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                },
                isAndroidNativeWebScreenshot: false,
                innerHeight: 500,
                isIOS: false,
            }
            const MOCKED_EXECUTOR = vi
                .fn()
                // getElementPositionDesktop for: getElementPositionTopWindow
                .mockResolvedValueOnce({
                    height: 20,
                    width: 375,
                    x: 12,
                    y: 34,
                })

            expect(
                await determineElementRectangles({
                    executor: MOCKED_EXECUTOR,
                    base64Image: IMAGE_STRING,
                    options,
                    element: 'element',
                }),
            ).toMatchSnapshot()
        })

        it('should throw an error when the element height is 0', async () => {
            const options = {
                isAndroid: false,
                devicePixelRatio: 2,
                deviceRectangles: {
                    homeBar: { y: 0, x: 0, width: 0, height: 0 },
                    statusBar: { y: 0, x: 0, width: 0, height: 0 },
                    screenSize: { height: 0, width: 0 },
                    statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
                    viewport: { y: 0, x: 0, width: 0, height: 0 },
                    bottomBar: { y: 0, x: 0, width: 0, height: 0 },
                    leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                    rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                },
                isAndroidNativeWebScreenshot: false,
                innerHeight: 500,
                isIOS: false,
            }
            const MOCKED_EXECUTOR = vi.fn().mockResolvedValueOnce({
                height: 0,
                width: 375,
                x: 12,
                y: 34,
            })

            try {
                await determineElementRectangles({
                    executor: MOCKED_EXECUTOR,
                    base64Image: IMAGE_STRING,
                    options,
                    element: { selector: '#elementID' },
                })
                // Fail test if above expression doesn't throw anything.
                expect(true).toBe(false)
            } catch (e: unknown) {
                expect((e as Error).message).toBe('The element, with selector "$(#elementID)",is not visible. The dimensions are 375x0')
            }
        })

        it('should throw an error when the element width is 0', async () => {
            const options = {
                isAndroid: false,
                devicePixelRatio: 2,
                deviceRectangles: {
                    homeBar: { y: 0, x: 0, width: 0, height: 0 },
                    statusBar: { y: 0, x: 0, width: 0, height: 0 },
                    screenSize: { height: 0, width: 0 },
                    statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
                    viewport: { y: 0, x: 0, width: 0, height: 0 },
                    bottomBar: { y: 0, x: 0, width: 0, height: 0 },
                    leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                    rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                },
                isAndroidNativeWebScreenshot: false,
                innerHeight: 500,
                isIOS: false,
            }
            const MOCKED_EXECUTOR = vi.fn().mockResolvedValueOnce({
                height: 375,
                width: 0,
                x: 12,
                y: 34,
            })

            try {
                await determineElementRectangles({
                    executor: MOCKED_EXECUTOR,
                    base64Image: IMAGE_STRING,
                    options,
                    element: { selector: '#elementID' },
                })
                // Fail test if above expression doesn't throw anything.
                expect(true).toBe(false)
            } catch (e: unknown) {
                expect((e as Error).message).toBe('The element, with selector "$(#elementID)",is not visible. The dimensions are 0x375')
            }
        })

        it('should throw an error when the element width is 0 and no element selector is provided', async () => {
            const options = {
                isAndroid: false,
                devicePixelRatio: 2,
                deviceRectangles: {
                    homeBar: { y: 0, x: 0, width: 0, height: 0 },
                    statusBar: { y: 0, x: 0, width: 0, height: 0 },
                    screenSize: { height: 0, width: 0 },
                    statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
                    viewport: { y: 0, x: 0, width: 0, height: 0 },
                    bottomBar: { y: 0, x: 0, width: 0, height: 0 },
                    leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                    rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
                },
                isAndroidNativeWebScreenshot: false,
                innerHeight: 500,
                isIOS: false,
            }
            const MOCKED_EXECUTOR = vi.fn().mockResolvedValueOnce({
                height: 375,
                width: 0,
                x: 12,
                y: 34,
            })

            try {
                await determineElementRectangles({
                    executor: MOCKED_EXECUTOR,
                    base64Image: IMAGE_STRING,
                    options,
                    element: {},
                })
                // Fail test if above expression doesn't throw anything.
                expect(true).toBe(false)
            } catch (e: unknown) {
                expect((e as Error).message).toBe('The element is not visible. The dimensions are 0x375')
            }
        })
    })

    describe('determineScreenRectangles', () => {
        it('should determine them for iOS', async () => {
            const options = {
                innerHeight: 553,
                innerWidth: 375,
                isAndroidNativeWebScreenshot: false,
                isAndroidChromeDriverScreenshot: false,
                isIOS: true,
                devicePixelRatio: 2,
                isLandscape: false,
            }

            expect(await determineScreenRectangles(IMAGE_STRING, options)).toMatchSnapshot()
        })

        it('should determine them for Android ChromeDriver', async () => {
            const options = {
                innerHeight: 553,
                innerWidth: 375,
                isAndroidNativeWebScreenshot: false,
                isAndroidChromeDriverScreenshot: true,
                isIOS: false,
                devicePixelRatio: 2,
                isLandscape: false,
            }

            expect(await determineScreenRectangles(IMAGE_STRING, options)).toMatchSnapshot()
        })

        it('should determine them for Android Native webscreenshot', async () => {
            const options = {
                innerHeight: 553,
                innerWidth: 375,
                isAndroidNativeWebScreenshot: true,
                isAndroidChromeDriverScreenshot: false,
                isIOS: false,
                devicePixelRatio: 2,
                isLandscape: false,
            }

            expect(await determineScreenRectangles(IMAGE_STRING, options)).toMatchSnapshot()
        })
    })

    describe('determineStatusAddressToolBarRectangles', () => {
        it('should determine the rectangles with a status and toolbar blockout', async () => {
            const options = {
                blockOutSideBar: true,
                blockOutStatusBar: true,
                blockOutToolBar: true,
                isAndroid: true,
                isAndroidNativeWebScreenshot: true,
                isMobile: true,
                isViewPortScreenshot: true,
            }
            const deviceRectangles =  {
                homeBar: { y: 0, x: 0, width: 0, height: 0 },
                statusBar: { y: 0, x: 0, width: 0, height: 0 },
                screenSize: { height: 0, width: 0 },
                statusBarAndAddressBar: { y: 0, x: 0, width: 1344, height: 320 },
                viewport: { y: 320, x: 0, width: 1344, height: 2601 },
                bottomBar: { y: 2921, x: 0, width: 1344, height: 71 },
                leftSidePadding: { y: 320, x: 0, width: 0, height: 2601 },
                rightSidePadding: { y: 320, x: 1344, width: 0, height: 2601 }
            }

            expect(await determineStatusAddressToolBarRectangles( { deviceRectangles, options })).toMatchSnapshot()
        })

        it('should determine the rectangles that there are no rectangles for this device', async () => {
            const options = {
                blockOutSideBar: false,
                blockOutStatusBar: false,
                blockOutToolBar: false,
                isAndroid: false,
                isAndroidNativeWebScreenshot: false,
                isMobile: true,
                isViewPortScreenshot: false,
            }
            const deviceRectangles =  {
                homeBar: { y: 0, x: 0, width: 0, height: 0 },
                statusBar: { y: 0, x: 0, width: 0, height: 0 },
                screenSize: { height: 0, width: 0 },
                statusBarAndAddressBar: { y: 0, x: 0, width: 1344, height: 320 },
                viewport: { y: 320, x: 0, width: 1344, height: 2601 },
                bottomBar: { y: 2921, x: 0, width: 1344, height: 71 },
                leftSidePadding: { y: 320, x: 0, width: 0, height: 2601 },
                rightSidePadding: { y: 320, x: 1344, width: 0, height: 2601 }
            }

            expect(await determineStatusAddressToolBarRectangles({ deviceRectangles, options })).toMatchSnapshot()
        })
    })
})
