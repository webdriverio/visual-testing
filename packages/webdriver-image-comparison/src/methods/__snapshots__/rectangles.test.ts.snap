// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rectangles > determineElementRectangles > should determine them for Android ChromeDriver 1`] = `
{
  "height": 20,
  "width": 375,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles > determineElementRectangles > should determine them for Android Native webscreenshot 1`] = `
{
  "height": 900,
  "width": 600,
  "x": 1200,
  "y": 630,
}
`;

exports[`rectangles > determineElementRectangles > should determine them for a desktop browser 1`] = `
{
  "height": 40,
  "width": 750,
  "x": 24,
  "y": 68,
}
`;

exports[`rectangles > determineElementRectangles > should determine them for iOS 1`] = `
{
  "height": 240,
  "width": 240,
  "x": 260,
  "y": 60,
}
`;

exports[`rectangles > determineScreenRectangles > should determine them for Android ChromeDriver 1`] = `
{
  "height": 1106,
  "width": 2732,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles > determineScreenRectangles > should determine them for Android Native webscreenshot 1`] = `
{
  "height": 1536,
  "width": 750,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles > determineScreenRectangles > should determine them for iOS 1`] = `
{
  "height": 1536,
  "width": 2732,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles > determineStatusAddressToolBarRectangles > should determine the rectangles that there are no rectangles for this device 1`] = `[]`;

exports[`rectangles > determineStatusAddressToolBarRectangles > should determine the rectangles with a status and toolbar blockout 1`] = `
[
  {
    "height": 320,
    "width": 1344,
    "x": 0,
    "y": 0,
  },
  {
    "height": 71,
    "width": 1344,
    "x": 0,
    "y": 2921,
  },
  {
    "height": 2601,
    "width": 0,
    "x": 0,
    "y": 320,
  },
  {
    "height": 2601,
    "width": 0,
    "x": 1344,
    "y": 320,
  },
]
`;
